src/raftstore/store/peer.rs

```diff
@@ -957,37 +957,26 @@ impl Peer {
 
     pub fn post_apply(
         &mut self,
-        res: &ApplyRes,
         groups: &mut HashSet<u64>,
-        store_stat: &mut StoreStat,
+        apply_state: RaftApplyState,
+        applied_index_term: u64,
+        merged: bool,
+        apply_metrics: &ApplyMetrics,
     ) {
         if self.is_applying_snapshot() {
             panic!("{} should not applying snapshot.", self.tag);
         }
 
-        let has_split = res.exec_res.iter().any(|e| match *e {
-            ExecResult::SplitRegion { .. } => true,
-            _ => false,
-        });
-
-        if !res.merged {
+        if !merged {
             self.raft_group
-                .advance_apply(res.apply_state.get_applied_index());
-        }
-        self.mut_store().apply_state = res.apply_state.clone();
-        self.mut_store().applied_index_term = res.applied_index_term;
-        self.peer_stat.written_keys += res.metrics.written_keys;
-        self.peer_stat.written_bytes += res.metrics.written_bytes;
-        store_stat.engine_total_bytes_written += res.metrics.written_bytes;
-        store_stat.engine_total_keys_written += res.metrics.written_keys;
-
-        let diff = if has_split {
-            self.delete_keys_hint = res.metrics.delete_keys_hint;
-            res.metrics.size_diff_hint
-        } else {
-            self.delete_keys_hint += res.metrics.delete_keys_hint;
-            self.size_diff_hint as i64 + res.metrics.size_diff_hint
-        };
+                .advance_apply(apply_state.get_applied_index());
+        }
+        self.mut_store().apply_state = apply_state;
+        self.mut_store().applied_index_term = applied_index_term;
+        self.peer_stat.written_keys += apply_metrics.written_keys;
+        self.peer_stat.written_bytes += apply_metrics.written_bytes;
+        self.delete_keys_hint += apply_metrics.delete_keys_hint;
+        let diff = self.size_diff_hint as i64 + apply_metrics.size_diff_hint;
```

Comment:
This may make split not happen in time. Ref https://github.com/pingcap/tikv/pull/3072/files#diff-eb8dcb6c9024d097a19d3a6de03b02b3R1620.